cli/deno_dir.rs

```diff
@@ -396,27 +404,234 @@ fn get_source_code_async(
+fn fetch_local_source(
+  deno_dir: &DenoDir,
+  module_name: &str,
+  filename: &str,
+  module_initial_source_name: Option<String>,
+) -> DenoResult<Option<ModuleMetaData>> {
+  let p = Path::new(&filename);
+  let source_metadata = get_source_code_metadata(&filename);
+  // If source metadata says that it would redirect elsewhere,
+  // (meaning that the source file might not exist; only .meta.json is present)
+  // Abort reading attempts to the cached source file and and follow the redirect.
+  if let Some(redirect_to) = source_metadata.redirect_to {
+    // e.g.
+    // module_name https://import-meta.now.sh/redirect.js
+    // filename /Users/kun/Library/Caches/deno/deps/https/import-meta.now.sh/redirect.js
+    // redirect_to https://import-meta.now.sh/sub/final1.js
+    // real_filename /Users/kun/Library/Caches/deno/deps/https/import-meta.now.sh/sub/final1.js
+    // real_module_name = https://import-meta.now.sh/sub/final1.js
+    let (real_module_name, real_filename) =
+      deno_dir.resolve_module(&redirect_to, ".")?;
+    let mut module_initial_source_name = module_initial_source_name;
+    // If this is the first redirect attempt,
+    // then module_initial_source_name should be None.
+    // In that case, use current module name as module_initial_source_name.
+    if module_initial_source_name.is_none() {
+      module_initial_source_name = Some(module_name.to_owned());
+    }
+    // Recurse.
+    return fetch_local_source(
+      deno_dir,
+      &real_module_name,
+      &real_filename,
+      module_initial_source_name,
+    );
+  }
+  // No redirect needed or end of redirects.
+  // We can try read the file
+  let source_code = match fs::read(p) {
+    Err(e) => {
+      if e.kind() == std::io::ErrorKind::NotFound {
+        return Ok(None);
+      } else {
+        return Err(e.into());
+      }
+    }
+    Ok(c) => c,
+  };
+  Ok(Some(ModuleMetaData {
+    module_name: module_name.to_string(),
+    module_redirect_source_name: module_initial_source_name,
+    filename: filename.to_string(),
+    media_type: map_content_type(
+      &p,
+      source_metadata.mime_type.as_ref().map(String::as_str),
+    ),
+    source_code,
+    maybe_output_code_filename: None,
+    maybe_output_code: None,
+    maybe_source_map_filename: None,
+    maybe_source_map: None,
+  }))
+}
+
+/// Asynchronously fetch remote source file specified by the URL `module_name`
+/// and write it to disk at `filename`.
+fn fetch_remote_source_async(
+  deno_dir: &DenoDir,
+  module_name: &str,
+  filename: &str,
+) -> impl Future<Item = Option<ModuleMetaData>, Error = DenoError> {
+  use crate::http_util::FetchOnceResult;
+  {
+    eprintln!("Downloading {}", module_name);
+  }
+
+  let filename = filename.to_owned();
+  let module_name = module_name.to_owned();
+
+  // We write a special ".meta.json" file into the `.deno/deps` directory along side the
+  // cached file, containing just the media type and possible redirect target.
```

Comment:
It would be better for extensibility to write `xxx.headers.json` which contained the HTTP headers of the response. Once loaded `headers.redirect_to()` could extract the location you're currently storing.

Reply:
SGTM. Renaming.